x/multisig/handlers.go

```diff
@@ -0,0 +1,174 @@
+	return res, nil
+}
+
+func (h CreateContractMsgHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (weave.DeliverResult, error) {
+	var res weave.DeliverResult
+	msg, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return res, err
+	}
+
+	contract := &Contract{
+		Sigs:                msg.Sigs,
+		ActivationThreshold: msg.ActivationThreshold,
+		AdminThreshold:      msg.AdminThreshold,
+	}
+
+	id := h.bucket.idSeq.NextVal(db)
+	obj := orm.NewSimpleObj(id, contract)
+	err = h.bucket.Save(db, obj)
+	if err != nil {
+		return res, err
+	}
+
+	res.Data = id
+	return res, nil
+}
+
+// validate does all common pre-processing between Check and Deliver
+func (h CreateContractMsgHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*CreateContractMsg, error) {
+	// Retrieve tx main signer in this context
+	sender := x.MainSigner(ctx, h.auth)
+	if sender == nil {
+		return nil, errors.ErrUnauthorized()
+	}
+
+	msg, err := tx.GetMsg()
+	if err != nil {
+		return nil, err
+	}
+
+	createContractMsg, ok := msg.(*CreateContractMsg)
+	if !ok {
+		return nil, errors.ErrUnknownTxType(msg)
+	}
+
+	err = createContractMsg.Validate()
+	if err != nil {
+		return nil, err
+	}
+
+	return createContractMsg, nil
+}
+
+type UpdateContractMsgHandler struct {
+	auth   x.Authenticator
+	bucket ContractBucket
+}
+
+var _ weave.Handler = CreateContractMsgHandler{}
+
+func (h UpdateContractMsgHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (weave.CheckResult, error) {
+	var res weave.CheckResult
+	_, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return res, err
+	}
+
+	res.GasAllocated = updateCost
+	return res, nil
+}
+
+func (h UpdateContractMsgHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (weave.DeliverResult, error) {
+	var res weave.DeliverResult
+	msg, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return res, err
+	}
+
+	contract := &Contract{
+		Sigs:                msg.Sigs,
+		ActivationThreshold: msg.ActivationThreshold,
+		AdminThreshold:      msg.AdminThreshold,
+	}
+
+	obj := orm.NewSimpleObj(msg.Id, contract)
+	err = h.bucket.Save(db, obj)
+	if err != nil {
+		return res, err
+	}
+
+	return res, nil
+}
+
+// validate does all common pre-processing between Check and Deliver
+func (h UpdateContractMsgHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*UpdateContractMsg, error) {
+	msg, err := tx.GetMsg()
+	if err != nil {
+		return nil, err
+	}
+
+	updateContractMsg, ok := msg.(*UpdateContractMsg)
+	if !ok {
+		return nil, errors.ErrUnknownTxType(msg)
+	}
+
+	err = updateContractMsg.Validate()
+	if err != nil {
+		return nil, err
+	}
+
+	// load contract
+	obj, err := h.bucket.Get(db, updateContractMsg.Id)
+	if err != nil {
+		return nil, err
+	}
+	if obj == nil || (obj != nil && obj.Value() == nil) {
+		return nil, ErrContractNotFound(updateContractMsg.Id)
+	}
+	contract := obj.Value().(*Contract)
+
+	// retrieve sigs
+	var sigs []weave.Address
```

Comment:
There really is no easier way to cast the type of an array in go?
I think I remember this issue and wishing it were easier, but maybe it is proper this way